### PR TITLE
Re-enable support for profiling/debugging interpreted/JITted code

### PR DESCRIPTION
**Pull request by amadio**
This feature, originally added in commit 22b1606f, was reverted to make the LLVM upgrade to version 13 easier. This commit adds back all functionality as it was just before the LLVM upgrade.

Edit: Unfortunately, still doesn't work with embedded debugging info:
```
epsftws build $ env CLING_DEBUG=1 EXTRA_CLING_ARGS='-gdwarf-5 -gembed-source' gdb -- bin/root.exe
GNU gdb (Gentoo 12.1 vanilla) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://bugs.gentoo.org/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from bin/root.exe...
(gdb) break f
Function "f" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (f) pending.
(gdb) run
Starting program: /srv/root/build/bin/root.exe 
warning: File "/srv/root/build/lib/libCore.so-gdb.py" auto-loading has been declined by your `auto-load safe-path' set to "$debugdir:$datadir/auto-load".
To enable execution of this file add
	add-auto-load-safe-path /srv/root/build/lib/libCore.so-gdb.py
line to your configuration file "/home/amadio/.config/gdb/gdbinit".
To completely disable this security protection add
	set auto-load safe-path /
line to your configuration file "/home/amadio/.config/gdb/gdbinit".
For more information about this security protection see the
"Auto-loading safe path" section in the GDB manual.  E.g., run from the shell:
	info "(gdb)Auto-loading safe path"
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib64/libthread_db.so.1".
[Detaching after vfork from child process 798483]
[Detaching after vfork from child process 800084]
[Detaching after vfork from child process 800087]
   ------------------------------------------------------------------
  | Welcome to ROOT 6.27/01                        https://root.cern |
  | (c) 1995-2022, The ROOT Team; conception: R. Brun, F. Rademakers |
  | Built for linuxx8664gcc on Dec 13 2022, 14:33:20                 |
  | From heads/cling-profile-debug@v6-25-02-3189-gf5fced388b         |
  | With c++ (Gentoo 12.2.1_p20221203 p3) 12.2.1 20221203            |
  | Try '.help'/'.?', '.demo', '.license', '.credits', '.quit'/'.q'  |
   ------------------------------------------------------------------

[Detaching after vfork from child process 800350]
root [0] double f(double x) { return x*x; }
root [1] double y = f(3.0);
Failure value returned from cantFail wrapped call
inconsistent use of embedded source
UNREACHABLE executed at /srv/root/src/root/interpreter/llvm/src/include/llvm/Support/Error.h:782!

Program received signal SIGABRT, Aborted.
__pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
44	      return INTERNAL_SYSCALL_ERROR_P (ret) ? INTERNAL_SYSCALL_ERRNO (ret) : 0;
(gdb) 

```
I will try with dwarf-5 for everything, let's see if that helps. At least GDB didn't crash this time.

### CODE REVIEW

**Comment by amadio** [comment]
Please clang-format...

**Comment by amadio** [comment]
Looks like my clang-format and the CI's clang-format disagree, so I will leave this as it is now. It seems that CI's clang-format wants shorter lines than before.

**Comment by Axel-Naumann** [comment]
The Windows error
```
[2022-12-13T12:03:53.345Z] clingInterpreter.lib(IncrementalJIT.obj) : error LNK2019: unresolved external symbol "class llvm::JITEventListener * __cdecl cling::createPerfJITEventListener(void)" (?createPerfJITEventListener@cling@@YAPAVJITEventListener@llvm@@XZ) referenced in function "private: virtual class llvm::Expected<class std::unique_ptr<class llvm::orc::ObjectLayer,struct std::default_delete<class llvm::orc::ObjectLayer> > > __thiscall std::_Func_impl_no_alloc<class <lambda_d2417646b7f0987897d54736770e7a35>,class llvm::Expected<class std::unique_ptr<class llvm::orc::ObjectLayer,struct std::default_delete<class llvm::orc::ObjectLayer> > >,class llvm::orc::ExecutionSession &,class llvm::Triple const &>::_Do_call(class llvm::orc::ExecutionSession &,class llvm::Triple const &)" (?_Do_call@?$_Func_impl_no_alloc@V<lambda_d2417646b7f0987897d54736770e7a35>@@V?$Expected@V?$unique_ptr@VObjectLayer@orc@llvm@@U?$default_delete@VObjectLayer@orc@llvm@@@std@@@std@@@llvm@@AAVExecutionSession@orc@3@ABVTriple@3@@std@@EAE?AV?$Expected@V?$unique_ptr@VObjectLayer@orc@llvm@@U?$default_delete@VObjectLayer@orc@llvm@@@std@@@std@@@llvm@@AAVExecutionSession@orc@4@ABVTriple@4@@Z) [C:\build\workspace\root-pullrequests-build\build\core\rootcling_stage1\rootcling_stage1.vcxproj]

[2022-12-13T12:03:53.345Z] C:\build\workspace\root-pullrequests-build\build\bin\rootcling_stage1.exe : fatal error LNK1120: 1 unresolved externals [C:\build\workspace\root-pullrequests-build\build\core\rootcling_stage1\rootcling_stage1.vcxproj]
```
seems both real and unnecessary :-)

**Comment by amadio** [comment]
I wonder how it was not there before, maybe I forgot some ifdefs...

**Comment by amadio** [comment]
Indeed, forgot an ifdef, should work now.

File: interpreter/cling/include/cling/Utils/Utils.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      const char* true_strs[] = {"1", "true", "on", "yes"};
+
+      if (!value)
+        return false;
+
+      for (auto str : true_strs)
+        if (strcasecmp(value, str) == 0)
+          return true;
+
+      return false;
```
Comment (Axel-Naumann) [inline interpreter/cling/include/cling/Utils/Utils.h:33]: This seems nicer for the 👀  ;-)
```suggestion
    static inline bool ConvertEnvValueToBool(llvm::StringRef value) {
      if (value.empty())
        return false;

      for (auto str : {"1", "true", "on", "yes"})
        if (value.equals_insensitive(str))
          return true;

      return false;
```

File: interpreter/cling/lib/Interpreter/CIFactory.cpp [hunk 2]
```diff
@@ -1323,6 +1331,16 @@ namespace {
     if(COpts.CUDAHost)
       argvCompile.push_back("--cuda-host-only");
 
+#ifdef __linux__
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/CIFactory.cpp:1334]: Why is this needed for setting argv but e.g. not for determining `profilingEnabled`? I don't get the plan here, for macOS and Windows: we change the argv flags for `debuggingEnabled` but not for `debuggingEnabled` - why?

File: interpreter/cling/lib/Interpreter/CIFactory.cpp [hunk 3]
```diff
@@ -1662,7 +1680,10 @@ namespace {
     // adjusted per transaction in IncrementalParser::codeGenTransaction().
     CGOpts.setInlining(CodeGenOptions::NormalInlining);
 
-    // CGOpts.setDebugInfo(clang::CodeGenOptions::FullDebugInfo);
+    // Add debugging info when debugging or profiling
+    if (debuggingEnabled || profilingEnabled)
+      CGOpts.setDebugInfo(codegenoptions::FullDebugInfo);
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/CIFactory.cpp:1685]: On all platforms? (This might be the thing to do, just asking because of the `#ifdef __linux__` above)

File: interpreter/cling/lib/Interpreter/PerfJITEventListener.cpp [hunk 4]
```diff
@@ -0,0 +1,129 @@
+//--------------------------------------------------------------------*- C++ -*-
+// CLING - the C++ LLVM-based InterpreterG :)
+// author: Guilherme Amadio <amadio@cern.ch>
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/PerfJITEventListener.cpp:3]: Where does this come from / what is it based on? It might be worth pointing out (in a comment) the source even if it's not a straight / partial copy, to tell us what to look at for future llvm upgrades.

File: interpreter/cling/lib/Interpreter/IncrementalJIT.cpp [hunk 5]
```diff
@@ -286,6 +287,9 @@ Error RTDynamicLibrarySearchGenerator::tryToGenerate(
 
 namespace cling {
 
+///\brief Creates JIT event listener to allow profiling of JITted code with perf
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/IncrementalJIT.cpp:290]: Could call out in a comment where this is implemented? E.g. just
```suggestion
///\brief Creates JIT event listener to allow profiling of JITted code with perf.
/// Implemented in `PerfJITEventListener.cpp`
```

File: interpreter/cling/include/cling/Utils/Utils.h (hunk 1 above)
Comment (Axel-Naumann) [reply interpreter/cling/include/cling/Utils/Utils.h:33]: I think I'd combine this with `getenv` (or rather [`llvm::sys::GetEnv`](https://llvm.org/doxygen/classllvm_1_1sys_1_1Process.html#a210453e575b2fc9a4a3e10bf8a9340fd)) in one function, to be called as `getEnvBool("CLING_PROFILE")` or similar.

**Review by Axel-Naumann — APPROVED** [review]
LGTM thanks! Couple of comments - please check whether they are worth addressing.

File: interpreter/cling/include/cling/Utils/Utils.h (hunk 1 above)
Comment (amadio) [reply interpreter/cling/include/cling/Utils/Utils.h:33]: Please see [this](https://github.com/root-project/root/pull/10890#discussion_r1013441471) discussion on `llvm::sys::GetEnv`.

Comment (amadio) [reply interpreter/cling/include/cling/Utils/Utils.h:33]: As a further comment, `StringRef` won't allow a `nullptr`, which I want to accept here as a false value (for when the environment variable is unset). I also thought I'd avoid the extra includes required for using a string type.

File: interpreter/cling/lib/Interpreter/CIFactory.cpp (hunk 2 above)
Comment (amadio) [reply interpreter/cling/lib/Interpreter/CIFactory.cpp:1334]: This part actually doesn't need to be wrapped around the ifdef, as we could enable the frame pointer on macOS, for example, even if the symbol map file is not created (as it's for `perf`, which is Linux-only.

Comment (amadio) [reply interpreter/cling/lib/Interpreter/CIFactory.cpp:1334]: The profiling feature is kinda perf-centric right now, which means that macOS and Windows need some work. Maybe we can try with Instruments on macOS and see if the frame-pointer helps and if some other event listener makes sense to help it see JITted code.

File: interpreter/cling/lib/Interpreter/CIFactory.cpp (hunk 3 above)
Comment (amadio) [reply interpreter/cling/lib/Interpreter/CIFactory.cpp:1685]: The debugging feature should work everywhere GDB works, so no ifdefs here.

File: interpreter/cling/lib/Interpreter/PerfJITEventListener.cpp (hunk 4 above)
Comment (amadio) [reply interpreter/cling/lib/Interpreter/PerfJITEventListener.cpp:3]: I looked at using `interpreter/llvm/src/lib/ExecutionEngine/PerfJITEvents/PerfJITEventListener.cpp` before I implemented this, as the JIT listener from LLVM did not work for me (it uses perf's [jitdump specification](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/tools/perf/Documentation/jitdump-specification.txt) rather than perf map files). I don't expect this file to need many changes in the future. I can add a comment, but the important thing is to just follow `llvm::JITEventListener`'s interfaces, as we derive from that class, so it's probably fine to leave as is.

Comment (amadio) [reply interpreter/cling/lib/Interpreter/PerfJITEventListener.cpp:3]: Also, there's already a [comment](https://github.com/root-project/root/pull/11882/files#diff-374f5def4dc7e413f703bac10548ff13d01bb20dd86c730ee5e57d881862c92eR69-R73) further down pointing to the LLVM listener...
